sprint_09

Gvozdkov opened the pull request.

Commit:
Изменен StatusBar с черного на белый

Commit:
1. Создан новый storyboard Profile.
2. Добавлены элементы: Avatar Image View, Name Label, Logout Button, Login Name Label, Description Label
3. Выставлены констрейнты согласно макету
4. Profile назначен экраном запуска
5. Добавлена связь с ProfileViewController

Commit:
В Main добавил UITabBarController и настроил связь между ImagesListViewController и ProfileViewController

Commit:
В Main к Images List View Controller добавлен UINavigationController и настроен согласно макету

Commit:
В Main  добавлен новый контроллер Single Image View Controller для просмотра отдельного изображения

Commit:
В Main настроен переход ячейки из Images List View Controller в Single Image View Controller

Commit:
Добавлена реализация при нажатия на ячейку с картинкой выполняется segue на экран Single Image View Controller и картинка отображается во весь экран

Commit:
В Single Image View Controller используется показ модального окна, т. е. картинка занимает весь экран и добавлена кнопка возврата

Commit:
Добавлен ScrollView на экран SingleImageViewController выставлены ему констрейнты, равные нулю со всех сторон.

Commit:
Реализована функциональность зума методом rescaleAndCenterImageInScrollView в SingleImageViewController, который изменяет скейл у scrollView так, чтобы картинка занимала по возможности весь экран, а её центр совпадал с центром экрана.

Commit:
В Single Image View Controller добавлена кнопка ShareButton при нажатие на которую происходит шеринг картинки. За функционал кнопки отвечает метод didTapShareButton.

Commit:
1. Main -> Profile Scene удалены все элементы, добавленные в Storyboard.
2. Все элементы были сверстаны в ProfileViewController без использования Storyboard.
3. Исправлены цвета в Assets.

Commit:
Исправлено отсутсвие логотипа при запуске приложения

ABashkirova commented on ImageFeed.xcodeproj/project.pbxproj:
```diff
@@ -39,6 +43,14 @@
 /* End PBXFrameworksBuildPhase section */
 
 /* Begin PBXGroup section */
+		E7017D8F2A10FA4A00D6EEA9 /* Profel */ = {
```
давай поправим опечатку

ABashkirova commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -7,16 +7,20 @@
     
     private let photosNames: [String] = Array(0..<20).map{ "\($0)" }
     
+    private let ShowSingleImageSegueIdentifier = "ShowSingleImage"
```
Предлагаю придерживаться общепринятого формата с именованием свойств/переменных/констант/функций и тд – в нижнем camelCase.

Не смотря на авторский стиль, все еще такую константу в общепринятых гайдах напишут с нижней s

ABashkirova commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -25,6 +29,18 @@ class ImagesListViewController: UIViewController {
+    
+    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
+        if segue.identifier == ShowSingleImageSegueIdentifier {
+            let viewController = segue.destination as! SingleImageViewController
+            let indexPath = sender as! IndexPath
+            let image = UIImage(named: photosNames[indexPath.row])
```
Здесь есть форс операции: это окей именно в этом месте, такое часто можно встретить даже в проектах в проде.

Однако я не рекомендую использовать этот способ в работах для портфолио: это не выглядит аккуратно, и понижает привлекательность работ перед работодателем.

ABashkirova commented on ImageFeed/Profel/ProfileViewController.swift:
```diff
@@ -0,0 +1,83 @@
+//  Created by Алексей Гвоздков on 14.05.2023.
+//
+
+import UIKit
+final class ProfileViewController: UIViewController {
```
```suggestion
import UIKit

final class ProfileViewController: UIViewController {
```

ABashkirova commented on ImageFeed/Profel/ProfileViewController.swift:
```diff
@@ -0,0 +1,83 @@
+
+import UIKit
+final class ProfileViewController: UIViewController {
+    private var viewProfileImage = UIImageView()
+    private var labelNameProfile = UILabel()
+    private var loginNameLabel = UILabel()
+    private var descriptionLabel = UILabel()
+    private var buttonLogout = UIButton()
+    
```
абсолютно все можно и нужно оставить константами. 
Это ссылочные типы, и любые настройки можно будет сделать далее в settingsViewController (не нужно переприсываивать. 

Еще можно пойти дальше и еще лучше: сделать настройку этих свойств с помощью хака с лейзи свойствами: 
```swift
private lazy var labelNameProfile: UILabel = {
 let label = UILabel()
 label.font = ...
 label.text = ...
 return label
}()
```

В методе settingsViewController останется только установить вьюшки на супервью, и настроить констрейнты

ABashkirova commented on ImageFeed/Profel/ProfileViewController.swift:
```diff
@@ -0,0 +1,83 @@
+        buttonLogout.tintColor = UIColor(named: "YP Red")
+        buttonLogout.translatesAutoresizingMaskIntoConstraints = false
+        view.addSubview(buttonLogout)
+        
+        
+        
```
давай договоримся, что между блоками не оставляем больше одной пустой строки

ABashkirova commented on ImageFeed/Profel/ProfileViewController.swift:
```diff
@@ -0,0 +1,83 @@
+        view.backgroundColor = UIColor(named: "YP Black")
+        
+        let imageProfile = UIImage(named: "Profile")
+        let imageButton = UIImage(named: "logoutButton")
```
тут инверсия в названии, общепринято в нейминге использовать для вьюшек такой порядок: "назначение"+тип. например, avatarImage, logoutButton и тд. 

Не настаиваю, так тебя тоже поймут.

ABashkirova commented on ImageFeed/SingleImageViewController/SingleImageViewController.swift:
```diff
@@ -0,0 +1,63 @@
+    }
+    
+    @IBOutlet private var imageView: UIImageView!
+    @IBOutlet weak var scrollView: UIScrollView!
```
вполне можно закрыть модификатором доступа.

ABashkirova commented on ImageFeed/Profel/ProfileViewController.swift:
```diff
@@ -0,0 +1,83 @@
+        
+        
+        NSLayoutConstraint.activate([
+            viewProfileImage.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 16),
+            viewProfileImage.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 16),
+            viewProfileImage.heightAnchor.constraint(equalToConstant: 70),
+            viewProfileImage.widthAnchor.constraint(equalToConstant: 70),
```
очень хорошо

ABashkirova commented on ImageFeed/Profel/ProfileViewController.swift:
```diff
@@ -0,0 +1,83 @@
+            viewProfileImage.heightAnchor.constraint(equalToConstant: 70),
+            viewProfileImage.widthAnchor.constraint(equalToConstant: 70),
+            
+            labelNameProfile.topAnchor.constraint(equalTo: viewProfileImage.bottomAnchor, constant: 8),
+            labelNameProfile.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 16),
+            
+            loginNameLabel.topAnchor.constraint(equalTo: labelNameProfile.bottomAnchor, constant: 8),
+            loginNameLabel.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 16),
+            
+            descriptionLabel.topAnchor.constraint(equalTo: loginNameLabel.bottomAnchor, constant: 8),
+            descriptionLabel.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 16),
```
хорошо! 

На мой взгляд, еще можно добавить симметричных trailing ограничений – на случай, если текст внутри будет слишком длинным, только с ограничением получится его обрезать/перенести на новую строку

ABashkirova commented on ImageFeed/Profel/ProfileViewController.swift:
```diff
@@ -0,0 +1,83 @@
+            descriptionLabel.topAnchor.constraint(equalTo: loginNameLabel.bottomAnchor, constant: 8),
+            descriptionLabel.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor, constant: 16),
+            
+            buttonLogout.centerYAnchor.constraint(equalTo: viewProfileImage.centerYAnchor),
+            buttonLogout.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor, constant: -26)
```
очень хорошо

IBDesignable commented on ImageFeed/SingleImageViewController/SingleImageViewController.swift:
```diff
@@ -0,0 +1,63 @@
+        scrollView.setContentOffset(CGPoint(x: x, y: y), animated: false)
+    }
+    
+    @IBAction private func didTapBackButton(_ sender: Any) {
+        dismiss(animated: true, completion: nil)
+    }
+    
+    @IBAction func didTapShareBotton(_ sender: Any) {
```
тоже можно закрыть приватным модификатором

IBDesignable commented on ImageFeed/SingleImageViewController/SingleImageViewController.swift:
```diff
@@ -0,0 +1,63 @@
+    }
+    
+    @IBAction func didTapShareBotton(_ sender: Any) {
+        let share = UIActivityViewController(
+        activityItems: [image],
+        applicationActivities: nil
+        )
```
Здесь поправить отступы:  control+i на выделенном участке

Commit:
Исправлено:
1. Изменен гейминг изображения с like_button_on на likeButtonOn и с like_button_off на likeButtonOff;
2. В ImagesListViewController переименована константа с ShowSingleImageSegueIdentifier на showSingleImageSegueIdentifier;
3. В ImagesListViewController все переменные с UI элементами стали лейзи с свойствами
4. Исправлена опечатка папки  с Profel на Profile

Commit:
В ImagesListViewController переменная с UIView перенесина с extension ProfileViewController ->  final class ProfileViewController: UIViewController
